### PR TITLE
No Auto Generated Themes

### PR DESCRIPTION
**Pull request by khacoder**
Issue #63
Verify that themes are not generated, by searching for strings contained in known auto generated packages.

### CODE REVIEW

**Comment by grappler** [comment]
This looks good. It will need an update to pass the automated tests.

**Comment by grappler** [comment]
I have rebased the branch so that it is up to date with the latest code and this get's travis to run it's latest checks. To update the branch locally run `git reset origin/feature/no-auto-generate --hard`

**Comment by grappler** [comment]
The PR is ready for the final review. I have simplified the checks to only check for the names. When  downloading the latest version of the generated themes I did not find the code that the check was looking for. These are not common words so there should not be any problem with false possitives.

File: WordPress/Sniffs/Theme/NoAutoGenerateSniff.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+		$tokens = $phpcsFile->getTokens();
+		$token  = $tokens[ $stackPtr ];
+
+		// No need to check an empty string.
+		if ( '' === trim( $token['content'] ) ) {
+			return;
+		}
+
+		$generated_themes = array(
```
Comment (jrfnl) [inline WordPress/Sniffs/Theme/NoAutoGenerateSniff.php]: This should be a class property, similar to the include/require sniff class property.

File: WordPress/Sniffs/Theme/NoAutoGenerateSniff.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+			'artisteer',
+			// Lubith.
+			'lubith',
+			// TemplateToaster.
+			'templatetoaster',
+			// WP Theme Generator.
+			'wpthemegenerator',
+		);
+
+		foreach ( $generated_themes as $generated_theme ) {
```
Comment (jrfnl) [inline WordPress/Sniffs/Theme/NoAutoGenerateSniff.php]: Similar to the favicon one, we could probably use a regex here, though as it's only four strings being sniffed for, it's not a breaking issue at the moment.
If/when the list gets longer, I would very much suggest changing it to a regex.

File: WordPress/Sniffs/Theme/NoAutoGenerateSniff.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+			if ( false !== strpos( strtolower( $token['content'] ), $generated_theme ) ) {
+				$phpcsFile->addError( 'Auto generated themes are not allowed in the theme directory.', $stackPtr, 'AutoGeneratedFound' );
```
Comment (jrfnl) [inline WordPress/Sniffs/Theme/NoAutoGenerateSniff.php]: Maybe add a `Found: %s` to the error sentence ? Or is this left out on purpose ?

File: WordPress/Tests/Theme/NoAutoGenerateUnitTest.inc [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+<p>Lubith</p>
+
+<?php
+// Lubith
+echo "Lubith";
+// Templatetoaster
+// wpthemegenerator.com
+/*
+ * @package templatetoaster
+ */
```
Comment (jrfnl) [inline WordPress/Tests/Theme/NoAutoGenerateUnitTest.inc]: As you are sniffing for both PHP as well as CSS, you will need a css example file too and deal with that in the unit test file.

File: WordPress/Sniffs/Theme/NoAutoGenerateSniff.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+		$generated_themes = array(
+			// Artisteer.
+			'artisteer',
+			// Lubith.
+			'lubith',
+			// TemplateToaster.
+			'templatetoaster',
+			// WP Theme Generator.
+			'wpthemegenerator',
```
Comment (jrfnl) [inline WordPress/Sniffs/Theme/NoAutoGenerateSniff.php]: Looks like the list has become a lot shorter than it originally was. Should the other snippets no longer be sniffed for ? How will they be dealt with ?

File: WordPress/Tests/Theme/NoAutoGenerateUnitTest.inc [hunk 6]
```diff
@@ -0,0 +1,11 @@
+
+<p>Lubith</p>
+
```
Comment (jrfnl) [inline WordPress/Tests/Theme/NoAutoGenerateUnitTest.inc]: Are these "real life" examples or made up examples ? I would very much prefer to see real life examples here to demonstrate what is being sniffed for and how.

Also, let's add some examples to make sure we don't get false positives.

File: WordPress/Sniffs/Theme/NoAutoGenerateSniff.php (hunk 5 above)
Comment (grappler) [inline WordPress/Sniffs/Theme/NoAutoGenerateSniff.php]: I mentioned in the comment above:

> When downloading the latest version of the generated themes I did not find the code that the check was looking for.

Comment (jrfnl) [inline WordPress/Sniffs/Theme/NoAutoGenerateSniff.php]: Ah, I misread that comment before.

Still, serial theme builders often don't download a new version of base code for each theme (take TGMPA for example... sigh...), so should we may be still check for code from at least a few versions back ?

Comment (jrfnl) [inline WordPress/Sniffs/Theme/NoAutoGenerateSniff.php]: You and the other members of the TRT are the better judge of that. I just wonder if relaxing the sniffs will not make some of them think "hang on, now we have a chance again...."

**Comment by grappler** [comment]
We need to add https://pinegrow.com/wordpress-theme-builder.html to the list.
Example of theme build with it: https://themes.svn.wordpress.org/dokumentasi-template/1.0/

**Comment by grappler** [comment]
@jrfnl I have made the requested changes. Thank you for pushing me use real world examples. Found a few bugs that way. Once I get the greenlight I can squash the commits.

**Comment by grappler** [comment]
I have updated file doc block. Do you have any ideas for test data to make sure we do not get false positives. I can't think of any. The keywords are unique enough that normal users should not be using them.

The only idea that I had was to split the keywords up so that they were normal words. e.g. `artist, them, template, toaster, pine, grow, theme, wp`

**Comment by jrfnl** [comment]
> Do you have any ideas for test data to make sure we do not get false positives. I can't think of any. The keywords are unique enough that normal users should not be using them.

Just some random thoughts on this:
1. CSS - add a header or two or ok themes (twentyseventeen and a non-automattic ok theme)
2. Add some lines containing things like: `artisteers`, `wpthemler`, `glubith`

Does that help ?

File: WordPress/Tests/Theme/NoAutoGenerateUnitTest.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	 * @return array <int line number> => <int number of errors>
+	 */
+	public function getErrorList( $testFile = '' ) {
+		switch ( $testFile ) {
+			case 'NoAutoGenerateUnitTest.inc':
+				return array(
+					2  => 1,
+					3  => 1,
+					6  => 1,
+					6  => 1,
```
Comment (jrfnl) [inline WordPress/Tests/Theme/NoAutoGenerateUnitTest.php]: Looks like there is a duplicate line here

**Comment by grappler** [comment]
Fixed the last bug. There is a chance for false positives if a theme uses any of the restricted words but the chances and reasons to do so it is very low.

**Pull request merged**